// safe-js/dist/dns.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.listServices = exports.getDns = exports.createLongName = exports.addService = exports.manifest = undefined;

var _isomorphicFetch = require('isomorphic-fetch');

var _isomorphicFetch2 = _interopRequireDefault(_isomorphicFetch);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
* Manifest for Beaker: 
* https://github.com/pfrazee/beaker/blob/master/doc/authoring-plugins.md#api-manifests
*/
var manifest = exports.manifest = {
  addService: 'promise',
  createLongName: 'promise',
  getDns: 'promise',
  listServices: 'promise'
};

// add service
var addService = exports.addService = function addService(token, longName, serviceName, isPathShared, serviceHomeDirPath) {
  var url = _utils.SERVER + 'dns';
  var payload = {
    method: 'PUT',
    headers: {
      Authorization: 'Bearer ' + token
    },
    data: {
      longName: longName,
      serviceName: serviceName,
      isPathShared: isPathShared,
      serviceHomeDirPath: serviceHomeDirPath
    }
  };
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200 && response.status !== 206) {
      throw new Error('SAFE addService failed with status ' + response.status + ' ' + response.statusText);
    }

    return response;
  });
};

var createLongName = exports.createLongName = function createLongName(token, longName) {
  var url = _utils.SERVER + 'dns/' + longName;
  var payload = {
    method: 'POST',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200 && response.status !== 206) {
      return Promise.reject('SAFE createLongName failed with status ' + response.status + ' ' + response.details);
    }

    if (response.ok) {
      return true;
    }
  });
};

// get dns list
var getDns = exports.getDns = function getDns(token) {
  var url = _utils.SERVER + 'dns';
  var payload = {
    method: 'GET',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error('SAFE getDns failed with status ' + response.status + ' ' + response.statusText);
    }

    return (0, _utils.parseResponse)(response);
  });
};

// get service
var listServices = exports.listServices = function listServices(token, longName) {
  var url = _utils.SERVER + 'dns/' + longName;
  var payload = {
    method: 'GET',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200 && response.status !== 206) {
      throw new Error('SAFE listServices failed with status ' + response.status + ' ' + response.statusText);
    }

    return (0, _utils.parseResponse)(response);
  });
};
